nemo/src/model/rule_model/term.rs

```diff
@@ -128,204 +43,393 @@ impl std::fmt::Display for Variable {
                 else {
-                    write!(f, "<{s}>")
+                    write!(f, "<{abstract_string}>")
                 }
             }
-            Term::Variable(term) => write!(f, "{term}"),
-            Term::NumericLiteral(term) => write!(f, "{term}"),
-            Term::StringLiteral(term) => write!(f, "\"{term}\""),
-            Term::RdfLiteral(term) => write!(f, "{term}"),
-            Term::Aggregate(aggregate) => write!(f, "{aggregate}"),
+            Constant::NumericLiteral(literal) => write!(f, "{}", literal),
+            Constant::StringLiteral(literal) => write!(f, "{}", literal),
+            Constant::RdfLiteral(literal) => write!(f, "{}", literal),
         }
     }
 }
 
-impl Term {
-    /// Check if the term is ground.
+/// Simple term that is either a constant or a variable
+#[derive(Debug, Eq, PartialEq, Clone, PartialOrd, Ord)]
+pub enum PrimitiveTerm {
+    /// A constant.
+    Constant(Constant),
+    /// A variable.
+    Variable(Variable),
+}
+
+impl From<Constant> for PrimitiveTerm {
+    fn from(value: Constant) -> Self {
+        Self::Constant(value)
+    }
+}
+
+impl PrimitiveTerm {
+    /// Return `true` if term is not a variable.
     pub fn is_ground(&self) -> bool {
-        matches!(
-            self,
-            Self::Constant(_) | Self::NumericLiteral(_) | Self::RdfLiteral(_)
-        )
+        !matches!(self, PrimitiveTerm::Variable(_))
     }
+}
 
-    /// Substitutes all occurrences of `var` for `subst`.
-    pub fn substitute_variable(&mut self, var: &Variable, subst: &Variable) {
-        debug_assert!(matches!(var, Variable::Universal(_)));
+impl Display for PrimitiveTerm {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            Term::Variable(v) => {
-                if v == var {
-                    *v = subst.clone()
-                }
-            }
-            Term::Aggregate(agg) => agg.substitute_variable(var, subst),
-            _ => {}
+            PrimitiveTerm::Constant(term) => write!(f, "{}", term),
+            PrimitiveTerm::Variable(term) => write!(f, "{}", term),
         }
     }
 }
 
-/// A numerical literal.
-#[derive(Debug, Eq, PartialEq, Copy, Clone, PartialOrd, Ord)]
-pub enum NumericLiteral {
-    /// An integer literal.
-    Integer(i64),
-    /// A decimal literal.
-    Decimal(i64, u64),
-    /// A double literal.
-    Double(Double),
+/// Represents an abstract logical function
+#[derive(Debug, Eq, PartialEq, Clone, PartialOrd, Ord)]
+pub struct AbstractFunction {
+    /// Name of the function
+    pub name: Identifier,
+    /// Its subterms
+    pub subterms: Vec<Term>,
 }
 
-impl std::fmt::Display for NumericLiteral {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// Possibly complex term that may occur within an [`super::Atom`]
+#[derive(Eq, PartialEq, Clone, PartialOrd, Ord)]
+pub enum Term {
```

Comment:
Again here I would strongly prefer something more structural like commented above, because otherwise I believe we risk spreading very long match statements with repetitive code.

Reply:
Yeah I can see that. I'll try to accommodate this